### DISCUSSION TITLE
[QST] can we first filter then join?

### DISCUSSION THREAD

**Discussion by JustPlay** [NVIDIA/spark-rapids#5389] [General] [answered]
https://github.com/NVIDIA/spark-rapids/blob/df00904433514930a2dbbdd8ae3ec9043dc3dc3d/shims/spark300/src/main/scala/com/nvidia/spark/rapids/shims/spark300/GpuHashJoin.scala#L257

https://github.com/NVIDIA/spark-rapids/blob/df00904433514930a2dbbdd8ae3ec9043dc3dc3d/shims/spark300/src/main/scala/com/nvidia/spark/rapids/shims/spark300/GpuHashJoin.scala#L258

https://github.com/NVIDIA/spark-rapids/blob/df00904433514930a2dbbdd8ae3ec9043dc3dc3d/shims/spark300/src/main/scala/com/nvidia/spark/rapids/shims/spark300/GpuHashJoin.scala#L268

**Comment by revans2** [comment, marked as answer]
Spark already does a filter push down as a part of join optimization.  If a conditional only involves one side of a join then it is pushed down before the join.  In other cases it cannot be pushed down so the filter happens after the join.  In fact filtering after the join only works for Inner joins because for all other joins you may need to evaluate the conditional for all combinations of matching keys to see if the condition will allow or exclude the row from being created.

We are working with cudf to be able to push the conditional into the join so we can reduce the amount of data that is materialized and increase conditional joins to include non-inner joins.  But that is not likely to show up until the 0.16 release of cudf at the earliest.
